README.md
# VPN Daemon

VPN Deemon is a simple daemon that provides a TCP socket API protected by TLS 
as an abstraction on top of the management port of (multiple) OpenVPN server 
process(es). The API exposes functionality to retrieve a list of connected VPN 
clients and also allows for disconnecting currently connected clients.

## Why?

On a VPN server we need to manage multiple OpenVPN processes. Each OpenVPN 
process exposes it management interface through a (TCP) socket. This works fine 
when the OpenVPN processes and the VPN controller run on the same machine. If 
both the controller and OpenVPN processes run on different hosts this is not 
secure as there is no TLS. Furthermore, it is inefficient, i.e. we have to 
query all OpenVPN management ports over the network for all hosts.

Currently, when using multiple hosts, one MUST have a secure channel between
the controller and node(s), which is something we do not want to require. A 
simple TLS channel protected by client certificate authentication over the open 
Internet should be enough...

This daemon will provide the exact same functionality as the current situation,
except the portal will talk to only one socket per VPN node, protected using 
TLS.

## How?

What we want to build is a simple daemon that runs on the same system as the 
OpenVPN processes and is reachable over a TCP socket protected by TLS. The 
daemon will then take care of contacting the OpenVPN processes through their 
local management ports and execute the commands. We want to make this 
"configuration-less", i.e. the daemon should require no additional 
configuration to make integrating it in the current system as easy as possible.

Currently there are two commands used over the OpenVPN management connection: 
`status` and `kill` where `status` returns a list of connected clients, and 
`kill` disconnects a client.

In a default installation our VPN server has two OpenVPN processes, so the 
daemon will need to talk to both OpenVPN processes. The portal can just talk to 
the daemon and issues a command there. The results will be merged by the 
daemon.

Furthermore, we can simplify the API used to retrieve the list of connected 
clients and disconnect clients. We will only expose what we explicitly use 
and need, nothing more.

## Architecture

                  .---------------------.
                  | Portal / Controller |
                  |                     |
                  '---------------------'
                           |
                           | Local/Remote TCP+TLS Socket
                           v
                  .----------------.
                  | Daemon         |
          .-------|                |-------.
          |       '----------------'       |
          |                                |
          |Local TCP Socket                |Local TCP Socket
          |                                |
          v                                v
    .----------------.               .----------------.
    | OpenVPN 1      |               | OpenVPN 2      |
    |                |               |                |
    '----------------'               '----------------'

## Benefits

The daemon will be written in Go, which can handle connections to the OpenVPN
management port concurrently. It doesn't have to perform the request one after 
the other as is currently the case. This may improve performance.

We can use TLS with the daemon and require TLS client certificate 
authentication. 

The parsing of the OpenVPN "legacy" protocol and merging of the 
information can be done by the daemon simplifying the implementation of the 
controller.

We can also begin to envision implementing other VPN protocols when we have
a control daemon, e.g. WireGuard. The daemon would need to have additional 
commands then, i.e. `setup` and `teardown`.

## API

### Command / Response

Currently 4 commands are implemented:

* `SET_PORTS`
* `DISCONNECT`
* `LIST`
* `QUIT`

The commands are given, some with parameters, and the response will be of the 
format:
    
    OK: n

Where `n` is the number of rows the response contains. This is an integer >= 0. 
See the examples below.

If a command is not supported, malformed, or a command fails, the response 
starts with `ERR`, e.g.:

    FOO
    ERR: NOT_SUPPORTED

### Setup

As we want to go for "zero configuration", we want the controller to specify 
which OpenVPN management ports we want to talk to.

    SET_PORTS 11940 11941

This works well for single profile VPN servers, but if there are multiple 
profiles involved, one has to specify them all in case of `DISCONNECT`, and 
a subset (just the ones for the profile one is interested in) when calling 
`LIST`.

### Disconnect 

`DISCONNECT` will disconnect the mentioned CN(s).

    DISCONNECT <CN_1> <CN_2> ... <CN_n>

Example:

    DISCONNECT 07d1ccc455a21c2d5ac6068d4af727ca
    
Response:

    OK: 0

### List

This will list all currently connected clients to the configured OpenVPN 
management ports. It exposes the management port as set by `SET_PORTS` to be 
able to link client to VPN process, `CN` and the IPv4 and IPv6 address assigned 
to the VPN client.

Example:

    LIST

Response:

    OK: 2
    11940 07d1ccc455a21c2d5ac6068d4af727ca 10.42.42.2 fd00:4242:4242:4242::1000
    11941 9b8acc27bec2d5beb06c78bcd464d042 10.132.193.3 fd0b:7113:df63:d03c::1001

### Quit

To close the connection:

    QUIT

## Build & Run

    $ go build -o _bin/vpn-daemon vpn-daemon/main.go

Or use the `Makefile`:

	$ make

## Run

    $ _bin/vpn-daemon

One can then `telnet` to port `41194`, and issue commands:

    $ telnet localhost 41194
    Trying ::1...
    Connected to localhost.
    Escape character is '^]'.
    SET_PORTS 11940 11941
    OK: 0
    DISCONNECT foo
    OK: 0
    QUIT

By default the daemon listens on `localhost:41194`. If you want to modify this
you can specify the `-listen` option to change this, e.g.:

    $ _bin/vpn-daemon -listen 192.168.122.1:41194

### TLS 

We use [vpn-ca](https://git.tuxed.net/LC/vpn-ca) to generate a CA:
    
    $ vpn-ca -init
    $ vpn-ca -server server
    $ vpn-ca -client client

If you want to enable TLS, i.e. require clients to connect over TLS, start 
the daemon with the `-enable-tls` flag, e.g.

    $ _bin/vpn-daemon -enable-tls

If you want to change the path where the CA, certificate and key are located, 
you can either patch ths file, or recompile the daemon with the flags 
`tlsCaPath`, `tlsCertPath` and `tlsKeyPath`, e.g.:

    $ go build -o _bin/vpn-daemon -ldflags="-X main.tlsCaPath=/path/to/ca.crt -X main.tlsCertPath=/path/to/server.crt -X main.tlsKeyPath=/path/to/server.key" vpn-daemon/main.go

In order to test the connection, you can use `openssl` and use it as a "telnet" 
to interact with the daemon:

    $ openssl s_client -connect 127.0.0.1:41194 -cert client.crt -key client.key -CAfile ca.crt

## Test

To run the test suite:

    $ make test


CHANGES.md
# ChangeLog

## 1.2.0 (2020-05-13)
- allow specifying full path to `tlsCaPath`, `tlsCertPath` and `tlsKeyPath` 
  isntead of folders allowing for more flexibility and easier packaging

## 1.1.1 (2020-05-01)
- update `Makefile` to support `install`

## 1.1.0 (2020-04-16)
- also return management port when using `LIST` command to be able to link 
  client connections to a particular OpenVPN process

## 1.0.1 (2020-04-08)
- fix parsing of OpenVPN status commando when no IP addresses are set 
  (issue #53)

## 1.0.0 (2019-11-18)
- initial release


vpn-daemon/main.go
/*
 * Copyright (c) 2019 François Kooman <fkooman@tuxed.net>
 *
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 *
 * The above copyright notice and this permission notice shall be included in all
 * copies or substantial portions of the Software.
 *
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 * SOFTWARE.
 */

package main

import (
	"bufio"
	"crypto/tls"
	"crypto/x509"
	"errors"
	"flag"
	"fmt"
	"io/ioutil"
	"net"
	"os"
	"regexp"
	"strconv"
	"strings"
	"sync"
	"time"
)

// the full path to the CA, certificate and key files
var (
	tlsCaPath   = "ca.crt"
	tlsCertPath = "server.crt"
	tlsKeyPath  = "server.key"
)

type vpnClientInfo struct {
	managementIntPort int
	commonName        string
	ipFour            string
	ipSix             string
}

func main() {
	var hostPort = flag.String("listen", "127.0.0.1:41194", "IP:port to listen on")
	var enableTls = flag.Bool("enable-tls", false, "enable TLS")
	flag.Usage = func() {
		flag.PrintDefaults()
	}
	flag.Parse()

	clientListener, err := getClientListener(*enableTls, *hostPort)
	fatalIfError(err)

	for {
		clientConnection, err := clientListener.Accept()
		if err != nil {
			fmt.Println(fmt.Sprintf("ERROR: %s", err))
			continue
		}

		go handleClientConnection(clientConnection)
	}
}

func getClientListener(enableTls bool, hostPort string) (net.Listener, error) {
	if enableTls {
		return tls.Listen("tcp", hostPort, getTlsConfig())
	}

	return net.Listen("tcp", hostPort)
}

func handleClientConnection(clientConnection net.Conn) {
	defer clientConnection.Close()

	managementIntPortList := []int{}
	setPortsRegExp := regexp.MustCompile(`^SET_PORTS [0-9]+( [0-9]+)*$`)
	disconnectRegExp := regexp.MustCompile(`^DISCONNECT [a-zA-Z0-9-.]+( [a-zA-Z0-9-.]+)*$`)
	writer := bufio.NewWriter(clientConnection)
	scanner := bufio.NewScanner(clientConnection)

	for scanner.Scan() {
		text := scanner.Text()
		fmt.Println(fmt.Sprintf("DEBUG: %s", text))

		// SET_PORTS
		if setPortsRegExp.MatchString(text) {
			parsedPortList, err := parseManagementPortList(strings.Fields(text)[1:])
			if err != nil {
				writer.WriteString(fmt.Sprintf("ERR: %s\n", err))
				writer.Flush()
				continue
			}

			managementIntPortList = parsedPortList
			writer.WriteString(fmt.Sprintf("OK: 0\n"))
			writer.Flush()
			continue
		}

		// DISCONNECT
		if disconnectRegExp.MatchString(text) {
			commonNameList := strings.Fields(text)[1:]
			// as "DISCONNECT" has no need to pass information back here, we
			// use a WaitGroup instead of a channel...
			var wg sync.WaitGroup
			for _, managementIntPort := range managementIntPortList {
				wg.Add(1)
				go disconnectClient(managementIntPort, commonNameList, &wg)
			}
			wg.Wait()
			writer.WriteString(fmt.Sprintf("OK: 0\n"))
			writer.Flush()
			continue
		}

		// LIST
		if text == "LIST" {
			vpnClientInfoChannel := make(chan []*vpnClientInfo, len(managementIntPortList))
			for _, managementIntPort := range managementIntPortList {
				go obtainStatus(managementIntPort, vpnClientInfoChannel)
			}

			vpnClientConnectionCount := 0
			vpnClientConnectionList := ""

			for range managementIntPortList {
				vpnClientInfoList := <-vpnClientInfoChannel
				for _, vpnClientInfo := range vpnClientInfoList {
					vpnClientConnectionCount++
					vpnClientConnectionList += fmt.Sprintf("%d %s %s %s\n", vpnClientInfo.managementIntPort, vpnClientInfo.commonName, vpnClientInfo.ipFour, vpnClientInfo.ipSix)
				}
			}

			writer.WriteString(fmt.Sprintf("OK: %d\n", vpnClientConnectionCount))
			writer.WriteString(vpnClientConnectionList)
			writer.Flush()
			continue
		}

		// QUIT
		if text == "QUIT" {
			writer.WriteString(fmt.Sprintf("OK: 0\n"))
			writer.Flush()
			return
		}

		writer.WriteString(fmt.Sprintf("ERR: NOT_SUPPORTED\n"))
		writer.Flush()
	}
}

func getManagementConnection(managementPort int) (net.Conn, error) {
	managementConnection, err := net.DialTimeout("tcp", fmt.Sprintf("127.0.0.1:%d", managementPort), time.Second*10)
	if err != nil {
		// problem establishing connection (timeout, closed, ...)
		fmt.Println(fmt.Sprintf("WARNING: %s", err))
		return nil, err
	}

	// make sure the connection does not hang forever reading/writing
	managementConnection.SetReadDeadline(time.Now().Add(time.Second * 3))

	return managementConnection, nil
}

func disconnectClient(managementPort int, commonNameList []string, wg *sync.WaitGroup) {
	defer wg.Done()
	managementConnection, err := getManagementConnection(managementPort)
	if err != nil {
		return
	}
	defer managementConnection.Close()

	managementPortScanner := bufio.NewScanner(managementConnection)
	// disconnect all CNs one-by-one
	for _, commonName := range commonNameList {
		fmt.Fprintf(managementConnection, "kill %s\n", commonName)
		for managementPortScanner.Scan() {
			text := managementPortScanner.Text()
			if strings.Index(text, "ERROR") == 0 || strings.Index(text, "SUCCESS") == 0 {
				// move on to next CN...
				break
			}
		}
	}
}

func obtainStatus(managementPort int, vpnClientInfoChannel chan []*vpnClientInfo) {
	managementConnection, err := getManagementConnection(managementPort)
	if err != nil {
		vpnClientInfoChannel <- []*vpnClientInfo{}
		return
	}
	defer managementConnection.Close()

	vpnClientInfoList := []*vpnClientInfo{}
	fmt.Fprintf(managementConnection, "status 2\n")
	managementPortScanner := bufio.NewScanner(managementConnection)
	for managementPortScanner.Scan() {
		text := managementPortScanner.Text()
		if strings.Index(text, "END") == 0 {
			break
		}
		if strings.Index(text, "CLIENT_LIST") == 0 {
			// HEADER,CLIENT_LIST,Common Name,Real Address,Virtual Address,
			//      Virtual IPv6 Address,Bytes Received,Bytes Sent,
			//      Connected Since,Connected Since (time_t),Username,
			//      Client ID,Peer ID
			strList := strings.Split(text, ",")
			if strList[1] != "UNDEF" && strList[3] != "" && strList[4] != "" {
				// only add clients with CN != "UNDEF" and IP addresses are not
				// empty strings...
				vpnClientInfoList = append(vpnClientInfoList, &vpnClientInfo{managementPort, strList[1], strList[3], strList[4]})
			}
		}
	}

	vpnClientInfoChannel <- vpnClientInfoList
}

func parseManagementPortList(managementStringPortList []string) ([]int, error) {
	managementIntPortList := []int{}
	for _, managementStringPort := range managementStringPortList {
		uintPort, err := strconv.ParseUint(managementStringPort, 10, 16)
		if err != nil || uintPort == 0 {
			return nil, errors.New("INVALID_PORT")
		}

		managementIntPortList = append(managementIntPortList, int(uintPort))
	}

	return managementIntPortList, nil
}

func getTlsConfig() *tls.Config {
	keyPair, err := tls.LoadX509KeyPair(tlsCertPath, tlsKeyPath)
	fatalIfError(err)

	caCertPem, err := ioutil.ReadFile(tlsCaPath)
	fatalIfError(err)

	trustedCaPool := x509.NewCertPool()
	if !trustedCaPool.AppendCertsFromPEM(caCertPem) {
		fatalIfError(errors.New("unable to add CA certificate to CA pool"))
	}

	return &tls.Config{
		Certificates: []tls.Certificate{keyPair},
		MinVersion:   tls.VersionTLS12,
		ClientAuth:   tls.RequireAndVerifyClientCert,
		ClientCAs:    trustedCaPool,
		CipherSuites: []uint16{tls.TLS_ECDHE_RSA_WITH_AES_256_GCM_SHA384},
	}
}

func fatalIfError(err error) {
	if err != nil {
		fmt.Printf("ERROR: %s\n", err)
		os.Exit(1)
	}
}
